devicetypes/smartthings/zigbee-window-shade.src/zigbee-window-shade.groovy

```diff
@@ -76,119 +76,170 @@ private List<Map> collectAttributes(Map descMap) {
 	if (descMap.additionalAttrs) {
 		descMaps.addAll(descMap.additionalAttrs)
 	}
-
 	return descMaps
 }
 
 // Parse incoming device messages to generate events
 def parse(String description) {
-    log.debug "description:- ${description}"
-    if (description?.startsWith("read attr -")) {
-        Map descMap = zigbee.parseDescriptionAsMap(description)
-        if (supportsLiftPercentage() && descMap?.clusterInt == CLUSTER_WINDOW_COVERING && descMap.value) {
-            log.debug "attr: ${descMap?.attrInt}, value: ${descMap?.value}, descValue: ${Integer.parseInt(descMap.value, 16)}, ${device.getDataValue("model")}"
-            List<Map> descMaps = collectAttributes(descMap)
-            def liftmap = descMaps.find { it.attrInt == ATTRIBUTE_POSITION_LIFT }
-            if (liftmap && liftmap.value) {
-                def newLevel = zigbee.convertHexToInt(liftmap.value)
-                levelEventHandler(newLevel)
-            }
-        } else if (!supportsLiftPercentage() && descMap?.clusterInt == zigbee.LEVEL_CONTROL_CLUSTER && descMap.value) {
-            def valueInt = Math.round((zigbee.convertHexToInt(descMap.value)) / 255 * 100)
-
-            levelEventHandler(valueInt)
-        }
-    }
+	log.debug "description:- ${description}"
+	if (description?.startsWith("read attr -")) {
+		Map descMap = zigbee.parseDescriptionAsMap(description)
+		if (isBindingTableMessage(description)) {
+			parseBindingTableMessage(description)
+		} else if (supportsLiftPercentage() && descMap?.clusterInt == CLUSTER_WINDOW_COVERING && descMap.value) {
+			log.debug "attr: ${descMap?.attrInt}, value: ${descMap?.value}, descValue: ${Integer.parseInt(descMap.value, 16)}, ${device.getDataValue("model")}"
+			List<Map> descMaps = collectAttributes(descMap)
+			def liftmap = descMaps.find { it.attrInt == ATTRIBUTE_POSITION_LIFT }
+			if (liftmap && liftmap.value) {
+				def newLevel = zigbee.convertHexToInt(liftmap.value)
+				if (shouldInvertLiftPercentage()) {
+					// some devices report % level of being closed (instead of % level of being opened)
+					// inverting that logic is needed here to avoid a code duplication
+					newLevel = 100 - newLevel
+				}
+				levelEventHandler(newLevel)
```

Comment:
Should we make sure to constrain `newLevel` to be between 0 and 100?

Reply:
We haven't done that in other places where we process level. It's _supposed_ to always be between 0 and 100